Classes/common/Indexing/CDTIndexManager.m

```diff
@@ -823,6 +759,132 @@ - (BOOL)isValidIndexName:(NSString *)indexName error:(NSError *__autoreleasing *
+{
+    NSString *dir = [datastore extensionDataFolder:kCDTExtensionName];
+    [[NSFileManager defaultManager] createDirectoryAtPath:dir
+                              withIntermediateDirectories:TRUE
+                                               attributes:nil
+                                                    error:nil];
+    NSString *filename = [NSString pathWithComponents:@[ dir, @"indexes.sqlite" ]];
+
+    FMDatabaseQueue *database = nil;
+    NSError *thisError = nil;
+    BOOL success = YES;
+
+    if (success) {
+        database = [[FMDatabaseQueue alloc] initWithPath:filename];
+
+        success = (database != nil);
+        if (!success) {
+            NSDictionary *userInfo = @{
+                NSLocalizedDescriptionKey :
+                    NSLocalizedString(@"Problem opening or creating database.", nil)
+            };
+            thisError = [NSError errorWithDomain:CDTIndexManagerErrorDomain
+                                            code:CDTIndexErrorSqlError
+                                        userInfo:userInfo];
+        }
+    }
+
+    if (success) {
+        id<CDTEncryptionKeyProvider> provider = [datastore copyEncryptionKeyProvider];
+
+        success = [CDTIndexManager configureDatabase:database withEncryptionKeyProvider:provider];
+
+        if (!success) {
+            NSDictionary *userInfo = @{
+                NSLocalizedDescriptionKey :
+                    NSLocalizedString(@"Problem configuring database with encryption key", nil)
+            };
+            thisError = [NSError errorWithDomain:CDTIndexManagerErrorDomain
+                                            code:CDTIndexErrorEncryptionKeyError
+                                        userInfo:userInfo];
+        }
+    }
+
+    if (success) {
+        success = [CDTIndexManager updateSchema:VERSION inDatabase:database];
+
+        if (!success) {
+            NSDictionary *userInfo = @{
+                NSLocalizedDescriptionKey :
+                    NSLocalizedString(@"Problem updating database schema.", nil)
+            };
+            thisError = [NSError errorWithDomain:CDTIndexManagerErrorDomain
+                                            code:CDTIndexErrorSqlError
+                                        userInfo:userInfo];
+        }
+    }
+
+    if (!success) {
+        database = nil;
+
+        if (error) {
+            *error = thisError;
+        }
+    }
+
+    return database;
+}
+
++ (BOOL)configureDatabase:(FMDatabaseQueue *)database
+withEncryptionKeyProvider:(id<CDTEncryptionKeyProvider>)provider
+{
+    __block BOOL success = YES;
+
+    NSString *key = [provider encryptionKey];
+    if (key) {
+        [database inDatabase:^(FMDatabase *db) {
+          success = [db setKey:key];
+        }];
+
+        if (!success) {
+            CDTLogError(CDTINDEX_LOG_CONTEXT, @"Key to cipher database not set");
+        }
+    }
+
+    return success;
+}
```

Comment:
But an index manager is created with an datastore. If the datastore is correctly initialised, this code will not fail. It will fail if the datastore is encrypted but the index is not, which means that the user copied in the folder an index created with a different datastore. Do we need to add defensive code for a situation like this?